docs/design/node_taint_eviction.md

```diff
@@ -0,0 +1,78 @@
+In Kubernetes it's common to use [taints](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration) to indicate potential problems with a node or planned maintenance.
+This information could be consumed by the operator to proactively replace Pods that are running on node with a given set of taints.
+
+## General Design Goals
+
+The goal of this design is to implement a mechanism that allows the operator to replace Pods that are running on nodes with a taint.
+
+## Current Implementation
+
+We don't have a current implementation of this and the operator is currently not able to access any node information.
+The closest implementation to this is the [Replace Failed ProcessGroups reconciler](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/controllers/replace_failed_process_groups.go), which could be extended for this.
+
+## Proposed Design
+
+We have to make sure that this feature can be enabled and disabled, e.g. if someone wants to run the operator inside a cluster without node access.
+The `AutomaticReplacementOptions` will be extended to support configuration of replacements for Pods that run on tainted nodes.
+The new field will be called `taints` and will contain a list, default would be empty, of taint configurations.
+That configuration will include the taint key, this has to be an exact match or a wildcard `*`, and the duration this taint must be present on the node.
+The `durationInSeconds` allow to specify how sensitive the operator should be to taints.
+If a wildcard is specified as key this will be used as default value, except there is an exact match defined for a taint key.
+
+```yaml
+...
+automaticReplacementOptions:
+  enabled: true
+  failureDetectionTimeSeconds: 7200
+  maxConcurrentReplacements: 1
+  enableNodeFailureDetection: true
+  taints:
+  - key: "example.org/maintenance"
+    durationInSeconds: 7200 # Ensure the taint is present for at least 2 hours before replacing Pods on a node with this taint
+  - key: "*" # The wildcard would allow to define a catch all configuration
+    durationInSeconds: 3600 # Ensure the taint is present for at least 1 hour before replacing Pods on a node with this taint
+```
+
+The actual implementation will extend the [ReplaceFailedProcessGroups](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/internal/replacements/replace_failed_process_groups.go) to handle those cases.
+
+```go
+    if !cluster.NodeFailureDetectionEnabled() {
+		return hasReplacements
+    }
+
+	// Only replace process groups if we can do a replacement. We could also track this as an independent metric.
+	if maxReplacements <= 0 {
+		return hasReplacements
+    }
+
+	pods, err := r.PodLifecycleManager.GetPods(ctx, r, cluster, internal.GetPodListOptions(cluster, "", "")...)
+    if err != nil {
+        return &requeue{curError: err}
+    }
+
+	nodeMap := internal.GetNodeMap(pods) // This will generate a map containing the node name as key and the pods as value.
```

Comment:
We either can make use of the `maxConcurrentReplacements` as a simple starting point or we could directly design it in a way that we always "drain" a node. Draining a node is probably more beneficial instead of only replacing a single Pod. We might want to add a `maxConcurrentNodeDrains` which would be a little bit more complex to track.

Reply:
A cluster only cares about how many concurrent pods can be put offline w/o losing availability. 
I think both replacement-on-failure and drain-on-taints are the same to cluster's availability.  
So a global `maxConcurrentReplacements` works better, imo.

Reply:
Based on https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1502/files#r1120462131 it's probably the best to use our normal mechanism and not add another setting.